### ISSUE TITLE
Execute local script over ssh

### ISSUE THREAD

**Issue by davegravy** [appleboy/ssh-action#186]
I have a bash script (.sh file) on my Github runner that I want to be executed on a remote host via SSH. I see that I can pass discrete commands using ssh-action, but how can an entire file be executed?

This would be similar to `ssh user@hostname < my-local-script.sh`

**Comment by devvspaces** [comment]
hi, i think you should be able to upload it on your github repository, so you can pull it on the server and run it

**Comment by devvspaces** [comment]
1. place .sh file on repo
2. ssh-action into the server
3. pull repo on the server
4. run .sh script

instead of step 3 i think you can also download the file to the server using  the absolute download link from github repo.
That would be better, instead of pulling an entire git repo

**Comment by brianjmurrell** [comment]
It's also really ugly to try to write more than a few lines of `sh` script in that `script:` tag, which has to be written as if it were all on a single line with `;` between commands, etc.  That script code cannot be linted for example.  If it cout be put into a file instead and the file executed with a `<` style redirection much of this problem goes away.

**Comment by brianjmurrell** [comment]
For those really wanting to be able to keep scripts in a separate file, there is this (admittedly ugly) hack:

```yaml
      - name: Read script contents
        id: script
        uses: andstor/file-reader-action@v1
        with:
          path: script/that/lives/in/workspace.sh
      - name: Execute local script
        uses: appleboy/ssh-action@master
        with:
          host: foo-240
          username: root
          key: ${{ secrets.POC }}
          script: ${{ steps.script.outputs.contents }}
```

**Comment by appleboy** [comment]
try [appleboy/scp-action](https://github.com/appleboy/scp-action) + appleboy/ssh-action

**Comment by brianjmurrell** [comment]
> try [appleboy/scp-action](https://github.com/appleboy/scp-action) + appleboy/ssh-action

But then I also have to have a job that cleans up those remotely **copied** scripts or leave litter on the remote machine.  Frankly I find this solution even worse than the [ugly hack I posted](https://github.com/appleboy/ssh-action/issues/186#issuecomment-1317513767) previously.

I would request that this ticket be reopened as it does not satisfy the OP's (or my) request to be able to execute the script remotely without all kinds of palaver to work around what should be straightforward and not require pre or post work of the actual step.

I'm not suggesting that it's a top priority or that you personally work on it, but AFAIC, it's still an incomplete feature request that could be worked on by anyone so it should remain open so that a potential contributor can see it and decide to pick it up.

**Comment by appleboy** [comment]
@brianjmurrell I got your point. I added the `enhancement` label now.

**Comment by demirtasdurmus** [comment]
Just to make things more clear, the purpose of this request is having the ability to achieve the same thing that can be done in bitbucket pipelines as follows: 

![image](https://github.com/user-attachments/assets/569a4643-e61b-435c-8c13-5b13db2cdf9e)

This ss shows the pipeline configuration, the mode is 'script' and the 'command' option accepts an sh file instead of a bunch of commands.

![image](https://github.com/user-attachments/assets/0d15af5d-00ab-439d-a3c2-bef47ec47596)

And this one is the actual script that is in the workspace and passed to the remote host in the pipeline. By this way, we can organize and build cleaner scripts in a separate file and run them inside our remote servers.

Thanks

**Comment by demirtasdurmus** [comment]
> The feature (#341) is released in [v1.2.0](https://github.com/appleboy/ssh-action/releases/tag/v1.2.0).
> 
> Set the `script_path` input parameter to a file path relative to the repository root and the action will read the file and run the commands.
> 
> Here's a sample:
> 
> ```yaml
> - name: file commands
>   uses: appleboy/ssh-action@v1.2.0
>   with:
>     host: ${{ secrets.HOST }}
>     username: ${{ secrets.USERNAME }}
>     key: ${{ secrets.KEY }}
>     port: ${{ secrets.PORT }}
>     script_path: scripts/script.sh 
> ```
> 
> @appleboy, I believe this issue is complete and can be closed!

Thanks. I think we might need some documentation on how to pass env variables to this script file, since it will be run in a different context.
